event-dispatcher/src/main/java/io/sk8s/event/dispatcher/EventDispatchingHandler.java

```diff
@@ -16,55 +16,264 @@
+		actualNbReplicas.put(functionName, deployment.getSpec().getReplicas());
+	}
+
+	@EventListener
+	public void onDeploymentUnregistered(ResourceDeletedEvent<Deployment> event) {
+		Deployment deployment = event.getResource();
+		String functionName = deployment.getMetadata().getName();
+		actualNbReplicas.remove(functionName);
+	}
+
+	@PreDestroy
+	public synchronized void tearDown() {
+		running = false;
+		this.notify();
+		kafkaConsumerMonitor.close();
+	}
+
+	private void updateWakeUpInterval() {
+		this.scalingFrequencyMs = this.functions.isEmpty() ? 0 : 100;
+	}
+
+	private class ScalingMonitor implements Runnable {
+
+		@Override
+		public void run() {
+			while (running) {
+				synchronized (EventDispatchingHandler.this) {
+
+					Map<KafkaConsumerMonitor.TopicGroup, List<KafkaConsumerMonitor.Offsets>> offsets = kafkaConsumerMonitor
+							.compute();
+					logOffsets(offsets);
+					Map<String, Long> lags = offsets.entrySet().stream()
+							.collect(Collectors.toMap(
+									e -> e.getKey().group,
+									e -> e.getValue().stream()
+											.mapToLong(KafkaConsumerMonitor.Offsets::getLag)
+											.max().getAsLong()));
+					functions.values().stream().forEach(
+							f -> {
+								int desired = computeDesiredNbReplicas(lags, f);
+								String name = f.getMetadata().getName();
+								Integer currentDesired = actualNbReplicas.computeIfAbsent(name, k -> desired);
+								System.out.println("Want " + desired + " for " + name);
+								if (desired < currentDesired) {
+									if (edgeInfos.computeIfAbsent(name,
+											n -> System.currentTimeMillis()) < System.currentTimeMillis()
+													- DOWN_EDGE_LINGER_PERIOD) {
+										deployer.deploy(f, desired);
+										actualNbReplicas.put(name, desired);
+										edgeInfos.remove(name);
+									}
+									else {
+										System.out.printf("Waiting another %dms before scaling down to %d for %s%n",
+												DOWN_EDGE_LINGER_PERIOD
+														- (System.currentTimeMillis() - edgeInfos.get(name)),
+												desired, name);
+									}
+								}
+								else if (desired > currentDesired) {
+									deployer.deploy(f, desired);
+									actualNbReplicas.put(name, desired);
+									edgeInfos.remove(name);
+								}
+								else {
+									edgeInfos.remove(name);
+								}
+							});
+
+					try {
+						EventDispatchingHandler.this.wait(scalingFrequencyMs);
+					}
+					catch (InterruptedException e) {
+						Thread.currentThread().interrupt();
+					}
+				}
+			}
+		}
+
+		/**
+		 * Compute the desired nb of replicas for a function. Each function defines 4 values:
+		 * <ul>
+		 * <li>minReplicas (>= 0, default 0)</li>
+		 * <li>maxReplicas (minReplicas <= maxReplicas <= nbPartitions, default nbPartitions)</li>
+		 * <li>l, the amount of lag required to trigger the first pod to appear, default 1</li>
+		 * <li>L, the amount of lag required to trigger all (maxReplicas) pods to appear. Default
+		 * 10.</li>
```

Comment:
I think the same value as "maxReplicas" would be a sensible default for the big L.

Reply:
That's unrelated. And remember the lag is understood _per partition_ (see computation below, which uses `max` as the aggregation function, not `sum`)

```java
Map<String, Long> lags = offsets.entrySet().stream()
	.collect(Collectors.toMap(
		e -> e.getKey().group,
		e -> e.getValue().stream()
			.mapToLong(KafkaConsumerMonitor.Offsets::getLag)
								.max().getAsLong()));
```

I still think there is no good, business agnostic default value. Can you explain your reasoning (incl if you think we change max to sum above)?

Reply:
Assume we're using `sum`. Are you voting for `bigL = max` because if lag (summed) == 10 with 10 partitions, then there is a chance that each partition has 1 msg pending?

Reply:
no, I understand that little l and big L are both being evaluated against the max lag seen on any one partition (aside: I actually think it would be smoother if you took the average and rounded up), but the value 10 seems arbitrary, whereas it could be proportional to the max number of replicas you would ever instantiate; i.e. the overall tolerance for lag depends on how much headroom is available for scaling. As an example, if I had 1000 partitions, wouldn't it be too aggressive to scale to 1000 instances just because I happened to see lag of 10 on the most-backed up partition during a single test at a 100ms interval?